### PR TITLE
DDG IA - Launchpad Project

### PR DESCRIPTION
**Pull request by Feral2k**
**What does your Instant Answer do?**

The IA searches and returns results for projects hosted on [Launchpad](https://launchpad.net/).

**What problem does your Instant Answer solve (Why is it better than organic links)?**

This provides results for projects hosted on Launchpad at a glance, including summaries of the projects and specifying the group that maintains the projects, rather than having to search through the Launchpad website itself.

**What is the data source for your Instant Answer? (Provide a link if possible)**

The reference documentation on the specific API used can be found here: https://launchpad.net/+apidoc/devel.html#projects

**Why did you choose this data source?**

It is the official Launchpad API to query and return results for packages hosted on that website.

**Are there any other alternative (better) data sources?**

I have not found any alternatives.

**What are some example queries that trigger this Instant Answer?**

"launchpad project elementary," "lp project bookworm"

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**

This IA will primarily benefit programmers and open-source enthusiasts to quickly find projects hosted on the website.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**

While not an exact match to an instant answer idea, it most closely resembles this one:
https://duck.co/ideas/idea/4740/launchpad-ppa-description-owner-and-package-list

**Which existing Instant Answers will this one supersede/overlap with?**

None.

**Are you having any problems? Do you need our help with anything?**

The instant answer appears to be functioning as I had intended, but I'm sure it needs more scrutiny and review.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**

I do not believe there needs to be any specific authorization or costs for the API. The terms of use for the API, so far as I can tell, are located here: https://help.launchpad.net/TermsofUse

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**

screenshot from 2015-07-31 09 49 15
screenshot from 2015-07-31 09 50 35
screenshot_2015-07-31-09-52-27

## Checklist
Please place an 'X' where appropriate.

```
[x] Added metadata and attribution information
[x] Wrote test file and added to t/ directory
[] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [] Google Chrome
        [] Firefox
        [] Opera
        [] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [x] Native Browser
        [] Google Chrome
        [] Opera
```
-----
IA Page: https://duck.co/ia/view/launchpad_project

### CODE REVIEW

**Comment by Feral2k** [comment]
Hmm, it seems that my test file has errors, if I'm understanding the Travis CI output correctly. I'm sure I made an obvious mistake somewhere. Let me know what I can do on my end to fix it.

**Comment by moollaza** [comment]
@Feral2k thanks a lot for contributing, we really appreciate it! We'll be sure to review this shortly and give you some feedback!

File: share/spice/launchpad_project/launchpad_project.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+(function (env) {
+    "use strict";        
+    env.ddg_spice_launchpad_project = function(api_result){
+
+        if (!api_result || !api_result.entries || api_result.error || api_result.total_size === 0) {
+            return Spice.failed('launchpad_project');
+        }
+        
+        //display "Launchpad Porject" if there is only one project returned
+        if (api_result.total_size === 1) {
```
Comment (moollaza) [inline share/spice/launchpad_project/launchpad_project.js]: We actually *just* made a function for this, which will be documented soon:

Here's an example:
```js
var amt = 1;
DDG.pluralize(amt, "leaf", "leaves"); //"leaf"

var amt = 3;
DDG.pluralize(amt, "leaf", "leaves") //"leaves"

var amt = 5;
DDG.pluralize(amt, "leaf") // "leafs" (when no plural form given, appends an "s" to singular form)
```

**Comment by moollaza** [comment]
@Feral2k this is looking pretty great. Judging by the docs, it looks like the API response contains a ton of info. Is there anything else we can show that would be useful to devs? Perhaps the date of creation or update? Programming language used? The project's icon?

@abeyang any thoughts on this one? I believe the API response contains all this info: https://launchpad.net/+apidoc/devel.html#project

**Comment by moollaza** [comment]
>Hmm, it seems that my test file has errors, if I'm understanding the Travis CI output correctly. I'm sure I made an obvious mistake somewhere. Let me know what I can do on my end to fix it.

Looks like we need to stop supressing the output from the tests! If you run the test locally, `prove -Ilib t/LaunchpadProject.t` you'll see it's failing because the tests contain "/query" in the URL instead of "/bookworm" and "/elementary"

File: t/LaunchpadProject.t [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+use strict;
+use warnings;
+use Test::More;
+use DDG::Test::Spice;
+
+ddg_spice_test(
+    [qw( DDG::Spice::LaunchpadProject)],
+    'launchpad project bookworm' => test_spice(
+        '/js/spice/launchpad_project/query',
```
Comment (moollaza) [inline t/LaunchpadProject.t]: this should be `/js/spice/launchpad_project/bookworm`

File: t/LaunchpadProject.t [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        call_type => 'include',
+        caller => 'DDG::Spice::LaunchpadProject'
+    ),
+    'lp project elementary' => test_spice(
+        '/js/spice/launchpad_project/query',
```
Comment (moollaza) [inline t/LaunchpadProject.t]: this should be `/js/spice/launchpad_project/elementary`

**Comment by Feral2k** [comment]
@moollaza Wow, what an embarrassing oversight on my part! Thanks so much! As for the other information, I completely agree. However, I believe many of them are links to other API results, and I am unsure of how or if I should wrap those API results in JSONP within the JavaScript file in order to display those elements. I'll try and see if there are other IAs that do this.

**Comment by Feral2k** [comment]
@moollaza I appear to have run into a bit of a problem when applying the DDG.pluralize function. Here is the code that I have replaced my previous code block with:
```
var amt = api_result.total_size;
var itemType = DDG.pluralize(amt, "Launchpad Project");
```
The code works as expected when returning a single result (the singular form) and two results (the plural form). But when returning 30 results, it displays nothing. Here are some test screenshots:

one-return
two-return
30-return

**Comment by abeyang** [comment]
@moollaza @Feral2k can you provide a sample of the returned API? The doc is a bit verbose, and I can't tell if all the fields are necessarily filled out (for example, do they all have logos or icons?).

**Comment by Feral2k** [comment]
@abeyang Here is a link to one of the API queries:

https://api.launchpad.net/devel/projects?ws.op=search&text=elementary&ws.size=30

An interesting thing to note is that the total_size field does not seem to be returned on that search query, but it is with others (such as 'bookworm' and 'slingshot'), which may be why the DDG.pluralize function didn't work. While there always appears to be an icon link field, in my early testing of the IA, I had the icon specified with the icon_link field and not all projects had an icon set, which is why I chose to get rid of it. Here is an example, where I have set the icon parameter under normalize as icon: item.icon_link:

icon_link

**Comment by abeyang** [comment]
@Feral2k yeah, no icons then. Let's go back to what you had previously, but for the footer, can you use `.bg-clr--grey-light` for the icon and `.tx-clr--grey-light` for the text?

**Comment by Feral2k** [comment]
@abeyang So just to make sure, are you looking for something like this?

icon-text-color

The icon seems to be a bit broken in night mode:

icon-text-color-night

**Comment by abeyang** [comment]
@Feral2k Yup. But what happened to the "user" icon? Your previous one looked more accurate (without the bounding box). Maybe remove the .bg* class to it and try applying `.tx-clr--grey-light` instead. We might also need to push the icon 1 or 2 pixels down.

**Comment by Feral2k** [comment]
@abeyang That fixed it! Unfortunately, I've been trying my best to move the icon down through the CSS file (in the function ```.tile--launchpad_project .ddgsi.ddgsi-user```) but to no avail. The ```padding``` and ```margin``` properties do nothing to it even though they effect a similar icon in the Github CSS file.

**Comment by abeyang** [comment]
Here's the fix for the icon:

``` css
.tileview__launchpad_project .ddgsi {
  position: relative;
  top: 1px;
}
```

In addition, can we tweak some font sizes? Please add `.tx--17` to the title, and `.tx--14` to both the content and footer. Thanks!

**Comment by Feral2k** [comment]
@abeyang Thanks so much! Here's how it looks so far using your recommendations:

new-look-light

And here's how it looks in dark:

new-look-dark

**Comment by abeyang** [comment]
:+1:

**Comment by Feral2k** [comment]
@abeyang Alright, I just pushed my latest changes to the repo. Your modifications should be all set. However, for some reason, my latest two commits aren't showing up here in this thread. Not sure why.

**Pull request merged**